### PR TITLE
Optimize `FluentTheme` updating

### PR DESCRIPTION
**Pull request by mischreiber**
### Platforms Impacted
- [x] iOS
- [x] visionOS
- [ ] macOS

### Description of changes

* Setting the shared `FluentTheme` now passes `nil` as the parent window into its notification, instead of iterating through all windows.
* Creating a new function, `FluentTheme.isApplicableThemeChange(_:for:)`, to capture the logic of determining whether a given `UIView` should update based on a theme notification (including the new `nil` logic used by `FluentTheme.shared`).
* Update all components that were doing manual notification observation to use this new API.
* Fix a bug in the Demo app re: per-window theme updates.

### Binary change

Total increase: 2,552 bytes
Total decrease: -43,408 bytes
| File | Before | After | Delta |
|------|-------:|------:|------:|
| Total | 31,061,536 bytes | 31,020,680 bytes | 🎉 -40,856 bytes |
<details>
<summary> Full breakdown </summary>

| File | Before | After | Delta |
|------|-------:|------:|------:|
| ColorProviding.o | 43,384 bytes | 45,560 bytes | ⚠️ 2,176 bytes |
| FocusRingView.o | 817,752 bytes | 818,128 bytes | ⚠️ 376 bytes |
| MSFNotification.o | 139,656 bytes | 139,528 bytes | 🎉 -128 bytes |
| MSFPersonaButtonCarousel.o | 32,696 bytes | 32,568 bytes | 🎉 -128 bytes |
| MSFPersonaButton.o | 37,840 bytes | 37,712 bytes | 🎉 -128 bytes |
| MSFHeadsUpDisplay.o | 34,072 bytes | 33,936 bytes | 🎉 -136 bytes |
| MSFAvatar.o | 27,552 bytes | 27,416 bytes | 🎉 -136 bytes |
| MSFAvatarGroup.o | 27,976 bytes | 27,840 bytes | 🎉 -136 bytes |
| MSFIndeterminateProgressBar.o | 28,632 bytes | 28,496 bytes | 🎉 -136 bytes |
| MSFActivityIndicator.o | 31,080 bytes | 30,944 bytes | 🎉 -136 bytes |
| MSFCardNudge.o | 36,496 bytes | 36,360 bytes | 🎉 -136 bytes |
| BadgeView.o | 625,216 bytes | 624,816 bytes | 🎉 -400 bytes |
| PageCardPresenterController.o | 146,992 bytes | 146,440 bytes | 🎉 -552 bytes |
| DatePickerController.o | 323,576 bytes | 322,952 bytes | 🎉 -624 bytes |
| MultilineCommandBar.o | 141,232 bytes | 140,552 bytes | 🎉 -680 bytes |
| DateTimePickerView.o | 207,936 bytes | 207,216 bytes | 🎉 -720 bytes |
| DateTimePickerController.o | 143,976 bytes | 142,936 bytes | 🎉 -1,040 bytes |
| PersonaListView.o | 187,120 bytes | 186,008 bytes | 🎉 -1,112 bytes |
| PopupMenuController.o | 383,968 bytes | 382,840 bytes | 🎉 -1,128 bytes |
| CalendarView.o | 71,248 bytes | 69,800 bytes | 🎉 -1,448 bytes |
| HUD.o | 259,640 bytes | 257,744 bytes | 🎉 -1,896 bytes |
| ControlHostingView.o | 65,944 bytes | 63,696 bytes | 🎉 -2,248 bytes |
| __.SYMDEF | 4,822,056 bytes | 4,818,744 bytes | 🎉 -3,312 bytes |
| CalendarViewWeekdayHeadingView.o | 86,040 bytes | 76,544 bytes | 🎉 -9,496 bytes |
| FluentTheme.o | 219,296 bytes | 201,744 bytes | 🎉 -17,552 bytes |
</details>

### Verification

Tested UIKit and SwiftUI controls, across multiple windows, to ensure standard expected behavior is maintained.

<details>
<summary>Visual Verification</summary>

https://github.com/microsoft/fluentui-apple/assets/4934719/54a81596-ad23-4dad-9c1a-29284856c99e

</details>

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [x] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [x] Objective-C exposure (provide it only if needed)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/fluentui-apple/pull/1976)

### CODE REVIEW

**Comment by harrieshin** [comment]
can we also update https://github.com/microsoft/fluentui-apple/wiki/Design-Tokens#fluenttheme  please?

**Comment by mischreiber** [comment]
[like]  Mike Schreiber (HE/HIM) reacted to your message:
________________________________

File: ios/FluentUI/Core/Theme/FluentTheme.swift [hunk 1]
```diff
@@ -92,19 +92,33 @@ public class FluentTheme: NSObject, ObservableObject {
...
+    ///
+    /// - Parameter notification: A `Notification` object that may be requesting a view update based on a theme change.
+    /// - Parameter view: The `UIView` instance that wants to determine whether to update.
+    ///
+    /// - Returns: `True` if the view should update, `false` otherwise.
+    @objc(isApplicableThemeChangeNotification:forView:)
+    public static func isApplicableThemeChange(_ notification: Notification,
+                                               for view: UIView) -> Bool {
+        // Do not update unless the notification's name is `.didChangeTheme`.
+        guard notification.name == .didChangeTheme else {
```
Comment (harrieshin) [inline ios/FluentUI/Core/Theme/FluentTheme.swift:109]: and our swift ui views are still relying that hostviews are listening into it? do swiftui environment trait still work as expected?

**Review by harrieshin — APPROVED** [review]

File: ios/FluentUI/Core/Theme/FluentTheme.swift (hunk 1 above)
Comment (mischreiber) [reply ios/FluentUI/Core/Theme/FluentTheme.swift:109]: Yes! The code in `ControlHostingView` bridges between this notification code (for when people set the theme directly on a `UIView`/`UIWindow`) and SwiftUI environment variables, where the theme is set using the `.fluentTheme(_:)` view modifier.

**Comment by mischreiber** [comment]
> can we also update https://github.com/microsoft/fluentui-apple/wiki/Design-Tokens#fluenttheme please?

Writing up new docs now. Also updated the comment block describing `FluentTheme.shared`.

**Pull request merged**